index.rst

```diff
@@ -13,17 +13,256 @@
+
+Implementation of important butler operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order for this proposal to be considered viable, we need to look carefully at how we would implement joint Registry/Datastore operations that we want to maintain consistency.
+Note that later sections in this technical note will expand upon these and ultimately take the form of prototype Python code, as further changes are added to the proposal.
+
+``Butler.put``
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
+
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
+
+4. If the butler has a Registry, either held directly (as in a "full butler" today) or as a client of a butler server, call ``Registry.insertDatasets`` with both the ``DatasetRef`` and the records returned by the Datastore.
+   Database transactions can be used to ensure that all tables in the Registry (including those for datastore records) are updated consistently or left unchanged.
+   If this operation fails, or the butler does not have a Registry (e.g. ``QuantumBackedButler``), the dataset is left in a valid state: it is in the Datastore, but not the Registry.
+   If the Datastore is backed by storage that requires signed URLs, this call also tells the server to transfer the artifacts from the staging area to permanent storage after checking for permissions.
+   This must happen before the database changes are committed.
+
+This has two major advantages over our current ``put`` implementation:
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
+
+- for a client/server butler, there is no alternation between object-store and http operations, just one and then the other, reducing latency (assuming the data ID has indeed been obtained in advance), and increasing the possiblity that the client Datastore can just be a regular ``FileDatastore``.
```

Comment:
I'm fairly certain it can't be a FileDatastore but I think it can be 99% FileDatastore with a small interface handling the URLs in a subclass. I don't see any issue with that given that we already know exactly how to switch in different datastores from configuration.

Reply:
Yup, after our discussion last Thursday I'm much, much less bothered by having a special Datastore, because I think it's special because it's talking to a special auth-protected storage endpoint, not because it's part of a client/server butler.

In fact, I wonder if "give me a signed URI for this dataset" should be a completely different server from the main butler server, just to establish that separation of concerns further.